src/dbnode/storage/index/aggregate_results.go

```diff
@@ -21,39 +21,106 @@
 package index
 
 import (
-	"fmt"
+	"math"
 	"sync"
 	"time"
 
-	"github.com/m3db/m3/src/m3ninx/doc"
+	"github.com/uber-go/tally"
+
 	"github.com/m3db/m3/src/m3ninx/index/segment/fst/encoding/docs"
 	"github.com/m3db/m3/src/x/ident"
+	"github.com/m3db/m3/src/x/instrument"
 	"github.com/m3db/m3/src/x/pool"
 )
 
-const missingDocumentFields = "invalid document fields: empty %s"
-
-// NB: emptyValues is an AggregateValues with no values, used for tracking
-// terms only rather than terms and values.
-var emptyValues = AggregateValues{hasValues: false}
-
 type aggregatedResults struct {
 	sync.RWMutex
 
 	nsID          ident.ID
 	aggregateOpts AggregateResultsOptions
 
 	resultsMap     *AggregateResultsMap
+	size           int
 	totalDocsCount int
 
 	idPool    ident.Pool
 	bytesPool pool.CheckedBytesPool
 
-	pool       AggregateResultsPool
-	valuesPool AggregateValuesPool
-
+	pool             AggregateResultsPool
+	valuesPool       AggregateValuesPool
 	encodedDocReader docs.EncodedDocumentReader
 	resultDuration   ResultDurations
+
+	iOpts instrument.Options
+}
+
+var _ AggregateUsageMetrics = (*usageMetrics)(nil)
+
+type usageMetrics struct {
+	total tally.Counter
+
+	totalTerms   tally.Counter
+	dedupedTerms tally.Counter
+
+	totalFields   tally.Counter
+	dedupedFields tally.Counter
+}
+
+func (m *usageMetrics) IncTotal(val int64) {
+	// NB: if metrics not set, to valid values, no-op.
+	if m.total != nil {
+		m.total.Inc(val)
+	}
+}
+
+func (m *usageMetrics) IncTotalTerms(val int64) {
+	// NB: if metrics not set, to valid values, no-op.
+	if m.totalTerms != nil {
+		m.totalTerms.Inc(val)
+	}
+}
+
+func (m *usageMetrics) IncDedupedTerms(val int64) {
+	// NB: if metrics not set, to valid values, no-op.
+	if m.dedupedTerms != nil {
+		m.dedupedTerms.Inc(val)
+	}
+}
+
+func (m *usageMetrics) IncTotalFields(val int64) {
+	// NB: if metrics not set, to valid values, no-op.
+	if m.totalFields != nil {
+		m.totalFields.Inc(val)
+	}
+}
+
+func (m *usageMetrics) IncDedupedFields(val int64) {
+	// NB: if metrics not set, to valid values, no-op.
+	if m.dedupedFields != nil {
+		m.dedupedFields.Inc(val)
+	}
+}
+
+// NewAggregateUsageMetrics builds a new aggregated usage metrics.
+func NewAggregateUsageMetrics(ns ident.ID, iOpts instrument.Options) AggregateUsageMetrics {
+	if ns == nil {
+		return &usageMetrics{}
```

Comment:
nit: Shouldn't this just return an error and the caller should return an error if construction fails?  

For tests we should probably just update callsites to always pass a non-nil ident.ID for the namespace.

Reply:
Yeah you're probably right, this felt really wonky